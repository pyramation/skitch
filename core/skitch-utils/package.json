{
  "name": "skitch-utils",
  "version": "0.16.58",
  "description": "",
  "author": "pyramation@gmail.com",
  "license": "MIT",
  "main": "./build/index.js",
  "types": "./build/types/index.d.ts",
  "scripts": {
    "build": "rm -rf ./build && tsc",
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "devDependencies": {
    "@types/jest": "21.1.0",
    "@types/node": "8.0.0",
    "dotenv": "5.0.1",
    "jest": "22.4.0",
    "ts-jest": "21.0.0",
    "ts-node": "5.0.0",
    "tslint": "5.4.3",
    "typescript": "2.7.1"
  },
  "dependencies": {
    "fuzzy": "0.1.3",
    "glob": "7.1.2",
    "mkdirp": "^0.5.1",
    "node-walkup": "^1.1.1",
    "pg": "^7.4.3",
    "pgsql-parser": "^0.5.8",
    "rimraf": "^2.6.2",
    "shelljs": "^0.8.2",
    "skitch-env": "^0.16.50",
    "skitch-ext-defaults": "0.0.1",
    "skitch-ext-utilities": "0.0.1",
    "skitch-ext-verify": "0.0.1",
    "skitch-template": "^0.16.58",
    "skitch-transform": "^0.16.50"
  },
  "jest": {
    "testURL": "http://localhost/",
    "transform": {
      ".(ts|tsx)": "<rootDir>/node_modules/ts-jest/preprocessor.js"
    },
    "testRegex": "(/__tests__/.*|\\.(test|spec))\\.(ts|tsx)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "coveragePathIgnorePatterns": [
      "/node_modules/",
      "/test/"
    ],
    "coverageThreshold": {
      "global": {
        "branches": 90,
        "functions": 95,
        "lines": 95,
        "statements": 95
      }
    }
  }
}
