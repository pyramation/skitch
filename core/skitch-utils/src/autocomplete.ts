import { filter } from 'fuzzy';

export const makeAutocompleteFunctionWithInput = (keys) => (answers: object, input: string) => {
  input = input || '';
  return new Promise(function(resolve) {
    setTimeout(function() {
      var fuzzyResult = filter(input, keys);
      resolve(
        fuzzyResult.map(function(el) {
          return el.original;
        })
      );
    }, 25);
  });
};
