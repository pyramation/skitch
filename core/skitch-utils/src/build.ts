import { listModules, getExtensionsAndModulesChanges } from './modules';
import { dirname, basename, resolve } from 'path';
import { skitchPath as sPath } from './paths';
import { readFileSync } from 'fs';

export const build = async argv => {
  const skitchPath = await sPath();
  const modules = await listModules();
  const modulesWithChanges = await getExtensionsAndModulesChanges('utils');

  const extensions = Object.keys(modules).map(key=>{
    const mod = modules[key];
    return {
      ...mod,
      sql: readFileSync(
        resolve(`${skitchPath}/${mod.path}/sql/${key}--${mod.version}.sql`)
      )
        .toString()
        .split('\n')
        .filter((l, i) => i !== 0)
        .join('\n')
    }
  });

  return extensions;
};
