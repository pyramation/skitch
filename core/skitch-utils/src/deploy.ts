import * as shell from 'shelljs';
import { resolve } from 'path';

import { listModules, getExtensionsAndModules } from './modules';

import { skitchPath } from './paths';

import { PGUSER, PGPASSWORD, PGHOST, PGPORT, PATH } from 'skitch-env';

const pg = require('pg');

// should we be parsing the plan file?
// currently assuming only extensions in control file...

export const deploy = async (name, database, opts) => {
  const path = await skitchPath();
  const modules = await listModules();
  const extensions = await getExtensionsAndModules(name);

  const pgPool = new pg.Pool({
    connectionString: `postgres://${PGUSER}:${PGPASSWORD}@${PGHOST}:${PGPORT}/${database}`
  });
  for (let n = 0; n < extensions.native.length; n++) {
    const extension = extensions.native[n];
    console.log(`CREATE EXTENSION IF NOT EXISTS "${extension}" CASCADE;`);
    await pgPool.query(
      `CREATE EXTENSION IF NOT EXISTS "${extension}" CASCADE;`
    );
  }
  pgPool.end();

  for (let n = 0; n < extensions.sqitch.length; n++) {
    const extension = extensions.sqitch[n];
    await deploy(extension, database, opts);
  }

  console.log(modules[name].path);
  console.log(`sqitch deploy db:pg:${database}`);
  shell.exec(`sqitch deploy db:pg:${database}`, {
    cwd: resolve(path, modules[name].path),
    env: {
      PGUSER,
      PGPASSWORD,
      PGHOST,
      PGPORT,
      PATH
    }
  });

  return extensions;
};
