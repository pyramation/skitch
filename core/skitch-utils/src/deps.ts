import { readFileSync, readFile } from 'fs';
import { basename, dirname, resolve, relative } from 'path';
import { sync as glob } from 'glob';
import { skitchPath } from './paths';
import { listModules } from './modules';

const makeKey = sqlmodule => '/deploy/' + sqlmodule + '.sql';

export const getDeps = async (packageDir) => {

  var external = [];

  var deps: { [type: string]: any } = {};
  var reg: { [type: string]: any } = {};

  // https://www.electricmonk.nl/log/2008/08/07/dependency-resolving-algorithm/
  function dep_resolve(
    sqlmodule: string,
    resolved: string[],
    unresolved: string[]
  ) {
    unresolved.push(sqlmodule);
    let edges = deps[makeKey(sqlmodule)];
    if (!edges) {
      if (/:/.test(sqlmodule)) {
        external.push(sqlmodule);
        edges = deps[sqlmodule] = [];
      } else {
        throw new Error(`no module ${sqlmodule}`);
      }
    }

    for (var i = 0; i < edges.length; i++) {
      var dep = edges[i];
      if (!resolved.includes(dep)) {
        if (unresolved.includes(dep)) {
          throw new Error(`Circular reference detected ${sqlmodule}, ${dep}`);
        }
        dep_resolve(dep, resolved, unresolved);
      }
    }

    resolved.push(sqlmodule);
    var index = unresolved.indexOf(sqlmodule);
    unresolved.splice(index);
  }

  var files = await glob(`${packageDir}/deploy/**/**.sql`);
  for (var i = 0; i < files.length; i++) {
    const data = readFileSync(files[i]);
    var lines = data.toString().split('\n');
    const key = '/' + relative(packageDir, files[i]);
    deps[key] = [];
    reg[key] = [];

    for (var j = 0; j < lines.length; j++) {
      var m = lines[j].match(/-- requires: (.*)/);
      if (m) {
        deps[key].push(m[1].trim());
      }

      // check only:
      var m2 = lines[j].match(/-- Deploy (.*) to pg/);
      if (m2) {
        if (key != makeKey(m2[1])) {
          throw new Error(
            'deployment script in wrong place or is named wrong internally' + m2
          );
        }
        reg[key].push(m2[1]);
      }
    }
  }

  let resolved: string[] = [];
  var unresolved: string[] = [];

  deps = Object.assign(
    {
      [makeKey('apps/index')]: Object.keys(deps)
        .filter(dep => dep.match(/^\/deploy\//))
        .map(dep => dep.replace(/^\/deploy\//, '').replace(/.sql$/, ''))
    },
    deps
  );

  dep_resolve('apps/index', resolved, unresolved);
  var index = resolved.indexOf('apps/index');
  resolved.splice(index);
  delete deps[makeKey('apps/index')];

  // move extensions up
  const extensions = resolved.filter(a => a.match(/^extensions/));
  const normalSql = resolved.filter(a => !a.match(/^extensions/));

  // resolved = useExtensions ? [...extensions, ...normalSql] : [...normalSql];
  resolved = [...extensions, ...normalSql];
  // resolved = [...normalSql];

  return {
    external,
    resolved,
    deps
  };

};
