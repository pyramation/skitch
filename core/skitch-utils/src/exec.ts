import { exec as shellExec } as shell from 'shelljs';
import { PGUSER, PGPASSWORD, PGHOST, PGPORT, PATH } from 'skitch-env';

export const execSync = (cmd, opts) => {
  shellExec(cmd, {
    env: {
      PGUSER,
      PGPASSWORD,
      PGHOST,
      PGPORT,
      PATH
    },
    ...opts
  });
};
