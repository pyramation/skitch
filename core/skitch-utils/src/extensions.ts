import { sync as glob } from 'glob';
import { sqitchPath as path } from './paths';
import { listModules } from './modules';
import { writeFileSync, readFileSync } from 'fs';
import { sluggify } from './utils';

export const getAvailableExtensions = async () => {
  let modules = await listModules();
  modules = Object.keys(modules).reduce(
    (m, v) => {
      if (m.includes(v)) return m;
      m.push(v);
      return m;
    },
    ['plpgsql', 'uuid-ossp', 'pgcrypto', 'plv8']
  );
  return modules;
};

export const getExtensionInfo = async () => {
  const sqitchPath = await path();
  const pkgPath = `${sqitchPath}/package.json`;
  const pkg = require(pkgPath);
  const extname = sluggify(pkg.name);
  const version = pkg.version;

  const Makefile = `${sqitchPath}/Makefile`;
  const controlFile = `${sqitchPath}/${extname}.control`;
  const sqlFile = `${extname}--${version}.sql`;
  return {
    extname,
    sqitchPath,
    version,
    Makefile,
    controlFile,
    sqlFile
  };
};

export const getInstalledExtensions = async () => {
  const info = await getExtensionInfo();

  let extensions;
  try {
    extensions = readFileSync(info.controlFile)
      .toString()
      .split('\n')
      .find(line => line.match(/^requires/))
      .split('=')[1]
      .split('\'')[1]
      .split(',')
      .map(a => a.trim());
  } catch (e) {
    throw new Error('missing requires from control files or bad syntax');
  }

  return extensions;
};

export const writeExtensionMakefile = async ({ path, extname, version }) => {
  writeFileSync(
    path,
    `EXTENSION = ${extname}
DATA = sql/${extname}--${version}.sql

PG_CONFIG = pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
  `
  );
};

export const writeExtensionControlFile = async ({
  path,
  extname,
  extensions,
  version
}) => {
  writeFileSync(
    path,
    `# ${extname} extension
comment = '${extname} extension'
default_version = '${version}'
module_pathname = '$libdir/${extname}'
requires = '${extensions.join(',')}'
relocatable = false
superuser = false
  `
  );
};

export const writeExtensions = async extensions => {
  const { controlFile: path, extname, version } = await getExtensionInfo();
  await writeExtensionControlFile({ path, extname, extensions, version });
};

export const writeExtensionsToEnv = async () => {
  const sqitchPath = await path();

  const controlFile = glob(`${sqitchPath}/*.control`);
  const envFile = glob(`${sqitchPath}/.env`);
  if (!controlFile || !controlFile.length) {
    throw new Error('no control file found!');
  }
  if (!envFile || !envFile.length) {
    throw new Error('no control file found!');
  }

  let extensions;
  try {
    extensions = readFileSync(controlFile[0])
      .toString()
      .split('\n')
      .find(line => line.match(/^requires/))
      .split('=')[1]
      .split('\'')[1]
      .split(',')
      .map(a => a.trim());
  } catch (e) {
    throw new Error('missing requires from control files or bad syntax');
  }

  let envs;
  try {
    envs = readFileSync(envFile[0])
      .toString()
      .split('\n')
      .reduce((m, line) => {
        line = (line || '').trim();
        if (/^#/.test(line)) return m;
        if (!line.length) return m;
        const parts = line.split('=');
        m[parts[0].trim()] = parts[1].trim();
        return m;
      }, {});
  } catch (e) {
    throw new Error('missing env files or bad syntax');
  }

  envs.PGEXTENSIONS = extensions.join(',');
  writeFileSync(
    envFile[0],
    Object.keys(envs).reduce((m, key) => {
      const value = envs[key];
      m = `${m}\n${key}=${value}`;
      return m;
    }, '')
  );
};
