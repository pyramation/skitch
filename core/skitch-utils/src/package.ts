import { sqitchPath as path } from './paths';
const parser = require('pgsql-parser');
import { resolve } from './resolve';
import { sync as mkdirp } from 'mkdirp';
import { relative } from 'path';
import { transformProps } from 'skitch-transform';
import { writeFileSync, readFileSync } from 'fs';
import  { sluggify } from './utils';

const noop = () => undefined;

export const cleanTree = (tree) => {
  return transformProps(tree, {
    stmt_len: noop,
    stmt_location: noop,
    location: noop
  });
};

export const packageModule = async (extension=true) => {
  const sqitchPath = await path();
  const sql = await resolve(sqitchPath);
  const pkgPath = `${sqitchPath}/package.json`;
  const pkg = require(pkgPath);
  const extname = sluggify(pkg.name);

  // sql
  try {
    const query = parser.parse(sql).query.reduce((m, stmt)=>{
      if (extension) {
        if (stmt.RawStmt.stmt.hasOwnProperty('TransactionStmt')) return m;
        if (stmt.RawStmt.stmt.hasOwnProperty('CreateExtensionStmt')) return m;
      }
      return [...m, stmt];
    }, []);
    const topLine = extension ? `\\echo Use "CREATE EXTENSION ${extname}" to load this file. \\quit\n` : '';
    const finalSql = parser.deparse(query);
    const tree1 = query;
    const tree2 = parser.parse(finalSql).query;
    const results = {
      sql: `${topLine}${finalSql}`
    };
    const diff = (JSON.stringify(cleanTree(tree1)) !== JSON.stringify(cleanTree(tree2)));
    if (diff) {
      results.diff = true;
      results.tree1 = JSON.stringify(cleanTree(tree1), null, 2);
      results.tree2 = JSON.stringify(cleanTree(tree2), null, 2);
    }

    return results;
  } catch (e) {
    console.error(e);
  }

};

export const writePackage = async (version, extension=true) => {
  const sqitchPath = await path();
  const pkgPath = `${sqitchPath}/package.json`;
  const pkg = require(pkgPath);
  const extname = sluggify(pkg.name);
  const makePath = `${sqitchPath}/Makefile`;
  const controlPath = `${sqitchPath}/${extname}.control`;
  const sqlFileName = `${extname}--${version}.sql`;

  const Makefile = readFileSync(makePath).toString();
  const control = readFileSync(controlPath).toString();

  const { sql, diff, tree1, tree2 } = await packageModule(extension);

  const outPath = extension ? `${sqitchPath}/sql` : `${sqitchPath}/out`;

  mkdirp(outPath);

  if (extension) {
    // control file
    writeFileSync(
      controlPath,
      control.replace(
        /default_version = '[0-9\.]+'/,
        `default_version = '${version}'`
      )
    );

    // package json
    writeFileSync(
      pkgPath,
      JSON.stringify(Object.assign({}, pkg, { version }), null, 2)
    );

    // makefile
    var regex = new RegExp(extname + '--[0-9.]+.sql');
    writeFileSync(makePath, Makefile.replace(regex, sqlFileName));
  }

  if (diff) {
    console.error(`DIFF exists! Careful. Check ${relative(sqitchPath, outPath)}/ folder...`);
    writeFileSync(`${outPath}/orig.${sqlFileName}.tree.json`, tree1);
    writeFileSync(`${outPath}/parsed.${sqlFileName}.tree.json`, tree2);
  }

  const writePath = `${outPath}/${sqlFileName}`;
  writeFileSync(writePath, sql);
  console.log(`${relative(sqitchPath, writePath)} written`);

};
