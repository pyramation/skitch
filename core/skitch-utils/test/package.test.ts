import { packageModule } from '../src/package';
process.env.SKITCH_PATH = __dirname + '/fixtures/skitch';
process.env.SQITCH_PATH = __dirname + '/fixtures/skitch/packages/secrets';

const clean = t =>
  t
    .split('\n')
    .map(a => a.trim())
    .filter(a => a)
    .join('\n');


describe('package', () => {
  it('creates an extension', async () => {
    const { sql } = await packageModule();
    expect(clean(sql)).toMatchSnapshot();
  });
});
