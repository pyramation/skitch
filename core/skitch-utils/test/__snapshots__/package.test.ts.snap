// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`package creates an extension 1`] = `
"\\\\echo Use \\"CREATE EXTENSION secrets\\" to load this file. \\\\quit
CREATE FUNCTION secretfunction (  ) RETURNS text AS $EOFCODE$
select * from generate_secret();
$EOFCODE$ LANGUAGE sql STABLE;"
`;
